device.go

```diff
@@ -229,37 +231,174 @@ func virtioBlkDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec,
+func vfioRebindDevice(bdf string) error {
+	fieldLogger := agentLog.WithField("guest-bdf", bdf)
+	fieldLogger.Debug("VFIO: Attempting to rebind")
+
+	sysPath := filepath.Join(sysBusPrefix, bdf)
+
+	// First we have to wait for the device to appear in the
+	// guest, since the hotplug can take a while.  We can't use
+	// the waiting logic in getDeviceName here, because the udev
+	// devices will only appear as we expect *after* we've
+	// rebound.
+	start := time.Now()
+	for {
+		_, err := os.Stat(sysPath)
+		if err == nil {
+			break
+		}
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("Error accessing %s: %s", sysPath, err)
+		}
+		if elapsed := time.Since(start); elapsed > vfioTimeout {
+			return fmt.Errorf("Timeout waiting for %s (%v)", sysPath, elapsed)
+		}
+		time.Sleep(vfioHotplugPollInterval)
+	}
+	fieldLogger.Debugf("VFIO: Found PCI device after %v", time.Since(start))
+
+	overridePath := filepath.Join(sysPath, "driver_override")
+	err := ioutil.WriteFile(overridePath, []byte("vfio-pci"), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", overridePath, err)
+	}
+
+	driverPath := filepath.Join(sysPath, "driver")
+	driver, err := os.Readlink(driverPath)
+	if os.IsNotExist(err) {
+		// This just means the device isn't bound to any driver
+		driver = ""
+	} else if err != nil {
+		return fmt.Errorf("Failed to readlink %s: %s", driverPath, err)
+	}
+
+	fieldLogger.Debugf("VFIO: Device previously bound to '%s'", driver)
+
+	if driver != "vfio-pci" && driver != "" {
+		fieldLogger.Debugf("VFIO: Unbinding from %s", driver)
+		unbindPath := filepath.Join(sysPath, bdf, "driver", "unbind")
+		err = ioutil.WriteFile(unbindPath, []byte(bdf), 0200)
+		if err != nil {
+			return fmt.Errorf("Failed to write %s: %s:", unbindPath, err)
+		}
+	}
+
+	err = ioutil.WriteFile(pciProbeFile, []byte(bdf), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", pciProbeFile, err)
+	}
+
+	fieldLogger.Info("VFIO: Rebind complete")
+	return nil
+}
+
+func vfioDeviceHandler(ctx context.Context, device pb.Device, spec *pb.Spec, s *sandbox, devIdx devIndex) error {
+	fieldLogger := agentLog.WithField("host-dev", device.ContainerPath)
+
+	var group string
+
+	for _, opt := range device.Options {
+		l := strings.Split(opt, "=")
+		if len(l) != 2 {
+			return fmt.Errorf("Malformed VFIO option \"%s\"", opt)
+		}
+
+		hostBdf := l[0]
+		guestPath := l[1]
+
+		guestAddr, err := getDevicePCIAddress(guestPath)
+		if err != nil {
+			return err
+		}
+
+		l = strings.Split(guestAddr, "/")
+		guestBdf := l[len(l)-1]
+
+		s.vfioDeviceMap[hostBdf] = guestBdf
+
+		err = vfioRebindDevice(guestBdf)
+		if err != nil {
+			return err
+		}
+
+		groupPath := filepath.Join(sysBusPrefix, guestBdf, "iommu_group")
+		groupLink, err := os.Readlink(groupPath)
+		if err != nil {
+			return fmt.Errorf("Failed to read %s: %s", groupPath, err)
+		}
+
+		if group == "" {
+			group = filepath.Base(groupLink)
+		} else if group != filepath.Base(groupLink) {
```

Comment:
Here, unlike for the `driver` case above, you are using the `Base`, which seems correct.

Reply:
Right.